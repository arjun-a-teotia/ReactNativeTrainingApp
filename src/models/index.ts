type User = {
  readonly id?: number;
  readonly name?: string;
  readonly email: string;
  readonly password: string;
};
type Pokemon = {
  readonly name?: string;
  readonly url: string;
};
type PokemonResource = {
  readonly count?: null | number;
  readonly next?: null | string;
  readonly previous: null | string;
  readonly results: Pokemon[];
};

export type {User, Pokemon, PokemonResource};
