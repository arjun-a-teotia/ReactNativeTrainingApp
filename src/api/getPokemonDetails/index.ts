import {Pokedex} from '../../models';

const getPokemonDetails = async (
  pokemonDetailsURL: string,
): Promise<Pokedex> => {
  const result = await fetch(pokemonDetailsURL);
  const {status} = result;

  if (status !== 200) {
    throw new Error(`Failed to get pokemon data (${status})`);
  }

  const response = await result.json();
  console.log('response', response);

  return response as Pokedex;
};

export {getPokemonDetails};
