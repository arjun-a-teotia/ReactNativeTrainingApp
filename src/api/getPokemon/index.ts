import {Pokemon, PokemonResource} from 'src/models';

const toModel = (response: PokemonResource): Array<Pokemon> => {
  return response.results;
};

const getPokemon = async (
  offset: number = 20,
  limit: number = 20,
): Promise<Pokemon[]> => {
  const baseUrl = 'https://pokeapi.co/api/v2';
  const url = `${baseUrl}/pokemon?offset=${offset}&limit=${limit}`;

  const result = await fetch(url);
  const {status} = result;

  if (status !== 200) {
    throw new Error(`Failed to get pokemon (${status})`);
  }

  const response = await result.json();

  return toModel(response);
};

export {getPokemon};
