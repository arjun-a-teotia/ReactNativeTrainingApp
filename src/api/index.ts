/* istanbul ignore file */
import {getPokemon} from './getPokemon';
import {getPokemonDetails} from './getPokemonDetails';

export {getPokemon, getPokemonDetails};
