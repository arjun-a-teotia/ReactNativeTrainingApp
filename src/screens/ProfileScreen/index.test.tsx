/**
 * @format
 */

import 'react-native';
import React from 'react';
import {ProfileScreen} from './index';

// Note: test renderer must be required after react-native.
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import analytics from '@react-native-firebase/analytics';
import {getPokemon} from '../..//api';
import {Pokemon} from '../../models';

const mockNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actual = jest.requireActual('@react-navigation/native');

  return {
    ...actual,
    useNavigation: () => ({
      navigate: mockNavigate,
      dispatch: mockNavigate,
    }),
  };
});

jest.mock('@react-native-firebase/app', () => {
  return () => ({
    app: jest.fn(),
  });
});
jest.mock('@react-native-firebase/auth', () => {
  return jest.fn().mockReturnValue({signOut: jest.fn()});
});
jest.mock('@react-native-firebase/analytics', () => {
  return jest.fn().mockReturnValue({logEvent: jest.fn()});
});
jest.useFakeTimers();
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.mock('src/api');

describe('Profile Screen Testing', () => {
  const mockGetPokemon = getPokemon as jest.Mock;

  const pokemons: readonly Pokemon[] = [
    {
      name: 'Leeds',
      url: 'www.google.com',
    },
  ];

  beforeEach(() => {
    // console.debug("mockGetPokemon", await mockGetPokemon());
    mockGetPokemon.mockResolvedValue(pokemons);
  });

  it('renders activity indicator (spinner) initially and then removes it', async () => {
    const {queryByTestId} = render(<ProfileScreen />);

    expect(queryByTestId('activity-indicator')).toBeTruthy();
    await waitFor(() =>
      expect(queryByTestId('activity-indicator')).toBeFalsy(),
    );
  });
  it('renders error message when fails to get pokemon list', async () => {
    const errorMessage = 'Failed to get pokemon list';

    mockGetPokemon.mockRejectedValue(new Error(errorMessage));

    const {getByText} = render(<ProfileScreen />);

    await waitFor(() => {
      expect(getByText(errorMessage)).toBeTruthy();
    });
  });
  it.each(pokemons)('renders pokemons list', async ({name}) => {
    const {getByTestId} = render(<ProfileScreen />);

    await waitFor(() => {
      expect(getByTestId(name)).toBeTruthy();
    });
  });
  it.each(pokemons)(
    'navigates to detail screen when user selects Pokemon "%j"',
    async pokemon => {
      const {getByTestId} = render(<ProfileScreen />);

      const {name} = pokemon;

      await waitFor(() => {
        expect(getByTestId(name)).toBeTruthy();
      });

      fireEvent.press(getByTestId(name));

      expect(mockNavigate).toHaveBeenCalledWith('PokemonDetailScreen', {
        pokemon,
      });
    },
  );

  it('navigates to login screen clicking logout button', async () => {
    const {getByTestId} = render(<ProfileScreen />);
    fireEvent.press(getByTestId('logout'));
    await waitFor(async () => {
      expect(analytics().logEvent).toBeCalledTimes(1);
    });
  });
});
