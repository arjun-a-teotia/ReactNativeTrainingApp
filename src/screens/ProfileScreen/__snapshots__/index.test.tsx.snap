// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Screen Testing renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Welcomeundefined
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "flex-end",
        "opacity": 1,
      }
    }
    testID="logout"
  >
    <Text
      style={
        Object {
          "color": "steelblue",
          "fontSize": 18,
          "marginTop": 150,
        }
      }
    >
      Logout
    </Text>
  </View>
</RNCSafeAreaView>
`;
